base/set.jl

```diff
@@ -447,22 +447,35 @@ false
 """
 function allunique(C)
     if haslength(C)
-        length(C) < 2 && return true
-        length(C) < 32 && return _indexed_allunique(collect(C))
+        len = length(C)
+        len < 2 && return true
+        len < 32 && return _indexed_allunique(collect(C), len)
+        return _hashed_allunique_haslength(C, len)
     end
-    return _hashed_allunique(C)
+    return _hashed_allunique_hasnolength(C)
 end
 
-function _hashed_allunique(C)
+function _hashed_allunique_hasnolength(C)
     seen = Set{eltype(C)}()
     x = iterate(C)
-    if haslength(C) && length(C) > 1000
-        for i in OneTo(1000)
+    while x !== nothing
+        v, s = x
+        in!(v, seen) && return false
+        x = iterate(C, s)
+    end
```

Comment:
This path exists twice. 

Could this not be coded more compactly by e.g. just defaulting to zero length `_hashed_allunique(C, len=0)` and only passing length if it is known? (Or defaulting to `len = nothing` if that's too obscure.)